docs: specifying files with arbitrary or no extension (#18539)
eslint/eslint: docs/src/use/configure/configuration-files.md

Patterns specified in `files` and `ignores` use [`minimatch`](https://www.npmjs.com/package/minimatch) syntax and are evaluated relative to the location of the `eslint.config.js` file. If using an alternate config file via the `--config` command line option, then all patterns are evaluated relative to the current working directory.
:::

> You can use a combination of `files` and `ignores` to determine which files should apply the configuration object and which should not. By default, ESLint matches `**/*.js`, `**/*.cjs`, and `**/*.mjs`. Because config objects that don't specify `files` or `ignores` apply to all files that have been matched by any other configuration object, only JavaScript files are linted by default. For example:

Reviewer: That last sentence doesn't seem correct to me. JavaScript linted by default isn't a result of config objects without `files` or `ignores`. Maybe this?
```suggestion
You can use a combination of `files` and `ignores` to determine which files should apply the configuration object and which should not. By default, ESLint matches `**/*.js`, `**/*.cjs`, and `**/*.mjs`. so those are always matched unless you explicitly exclude them using `ignores`. Because config objects that don't specify `files` or `ignores` apply to all files that have been matched by any other configuration object, they will apply to all JavaScript files. For example:
```

Author: Yeah, actually it's a result of the default patterns being what they are and also the fact that config objects without `files` or `ignores` apply to all linted files. I will elaborate.

Non-global `ignores` patterns can only match file names. A pattern like `"dir-to-exclude/"` will not ignore anything. To ignore everything in a particular directory, a pattern like `"dir-to-exclude/**"` should be used instead.

> If `ignores` is used without `files` and there are other keys (such as `rules`), then the configuration object applies to all linted files except the ones specified in `ignores`, for example:

Reviewer: ```suggestion
If `ignores` is used without `files` and there are other keys (such as `rules`), then the configuration object applies to all linted files except the ones excluded by `ignores`, for example:
```

];
```

This configuration object applies to all JavaScript files except those ending with `.config.js`. Effectively, this is like having `files` set to `**/*`. In general, it's a good idea to always include `files` if you are specifying `ignores`.

Note that negated `ignores` patterns do not cause any matching files to be linted automatically.

> ESLint only lints files that are matched either by default or by a `files` pattern not ending with `/*` or `/**`.

Reviewer: It seems like this is missing some context. Do you mean negated `ignores` without a `files`?

::: tip
Use the [config inspector](https://github.com/eslint/config-inspector) (`--inspect-config` in the CLI) to test which config objects apply to a specific file.
:::

#### Specifying files with arbitrary extensions

> To lint files with other extensions than the default `.js`, `.cjs` and `.mjs`, include them in `files` with a glob pattern.
> Any pattern that doesn't end with `/*` or `/**` will work.
> For example, to lint TypeScript files with `.ts`, `.cts` and `.mts` extensions:

Reviewer: ```suggestion
To lint files with extensions other than the default `.js`, `.cjs` and `.mjs`, include them in `files` with a global pattern in the format of `"**/*.extension"`. (Any pattern that doesn't end with `/*` or `/**` will work.)
For example, to lint TypeScript files with `.ts`, `.cts` and `.mts` extensions, you would specify a configuration object like this:
```

```js
// eslint.config.js
export default [
    {
        files: [
            "**/*.ts",
            "**/*.cts",
            "**.*.mts"
        ]
    },
    // ...other config
];
```

#### Specifying files without extension

> Files without extension can be matched with the pattern `!(*.*)`. For example:

Reviewer: ```suggestion
Files without an extension can be matched with the pattern `!(*.*)`. For example:
```

```js
// eslint.config.js
export default [
    {
        files: ["**/!(*.*)"]
    },
    // ...other config
];
```

The above config lints files without extension besides the default `.js`, `.cjs` and `.mjs` extensions in all directories.

> Note that filenames starting with a dot are considered to have an extension.

Reviewer: To clarify:
```suggestion
::: tip
Filenames starting with a dot, such as `.gitignore`, are considered to have only an extension without a base name. In the case of `.gitignore`, the extension is `gitignore`, so the file matches the pattern `"**/.gitignore"` but not `"**/*.gitignore"`.
:::
```

];
```

This configuration object applies to all JavaScript files except those ending with `.config.js`. Effectively, this is like having `files` set to `**/*`. In general, it's a good idea to always include `files` if you are specifying `ignores`.

Note that negated `ignores` patterns do not cause any matching files to be linted automatically.
ESLint only lints files that are matched either by default or by a `files` pattern not ending with `/*` or `/**`.

::: tip
Use the [config inspector](https://github.com/eslint/config-inspector) (`--inspect-config` in the CLI) to test which config objects apply to a specific file.
:::

#### Specifying files with arbitrary extensions

To lint files with other extensions than the default `.js`, `.cjs` and `.mjs`, include them in `files` with a glob pattern.
Any pattern that doesn't end with `/*` or `/**` will work.
For example, to lint TypeScript files with `.ts`, `.cts` and `.mts` extensions:

```js
// eslint.config.js
export default [
    {
        files: [
            "**/*.ts",
            "**/*.cts",
            "**.*.mts"
        ]
    },
    // ...other config
];
```

#### Specifying files without extension

> No simple glob pattern can be used to target only files without extension.

Reviewer 2: A pattern like `**/!(*.*)` would match only files without extension I believe.
